### PR TITLE
feat: other players loading progress in loading screen [MTT-2239]

### PR DESCRIPTION
**Pull request by LPLafontaineB**
### Description (*)

This PR modifies the loading screen so that it can also display the loading progress of other players. To accomplish this, I added a simple player-owned NetworkObject that contains an owner-writtable NetworkVariable to hold the player's current progress. These are spawned by LoadingProgressManager on the server and kept in a map on all clients, so that a client can update its progress and read the progress of the other clients.
### Issue Number(s) (*)

Fixes issue(s): [MTT-2239](https://jira.unity3d.com/browse/MTT-2239)
### Manual testing scenarios

1. Start a game as a host
2. Connect clients to the session
3. Complete character selection and move to the next scene
4. Disconnect a client
5. Connect a new client
6. Go to Post-game
7. Leave the game session
The loading screen should be properly displayed all those times, and should display the progress of other connected players. The progress bars should update when characters disconnect and when new ones connect.
### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/ClientLoadingScreen.cs [hunk 1]
```diff
@@ -63,26 +82,88 @@ public void StopLoadingScreen()
...
+            // clear map
+            m_ClientIdToProgressBarsIndex.Clear();
+
+            // deactivate all other players' progress bars
+            foreach (var progressBar in m_OtherPlayersProgressBars)
+            {
+                progressBar.gameObject.SetActive(false);
+            }
+
+            if (IsSpawned && !NetworkManager.ShutdownInProgress)
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/ClientLoadingScreen.cs:118]: check if loading manager is spawned and remove NetworkBehaviour

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/ClientLoadingScreen.cs [hunk 2]
```diff
@@ -63,26 +82,88 @@ public void StopLoadingScreen()
...
+                            m_ClientIdToProgressBarsIndex[clientId] = m_ClientIdToProgressBarsIndex.Count;
+                            // set progress bar to 0 if initializing, else set it to its last known value
+                            m_OtherPlayersProgressBars[m_ClientIdToProgressBarsIndex[clientId]].value = isInitializing ? 0 : progress.Value;
+                            progress.OnValueChanged += (value, newValue) =>
+                                m_OtherPlayersProgressBars[m_ClientIdToProgressBarsIndex[clientId]].value = newValue;
+                            m_OtherPlayersProgressBars[m_ClientIdToProgressBarsIndex[clientId]].gameObject.SetActive(true);
+                        }
+                        else
+                        {
+                            Debug.LogError("There are not enough progress bars to track the progress of all the players.");
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/ClientLoadingScreen.cs:146]: Throw exception if this is a sanity checks

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/NetworkedLoadingProgressTracker.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+using Unity.Netcode;
+
+namespace Unity.Multiplayer.Samples.Utilities
+{
+    /// <summary>
+    /// Simple object that keeps track of the scene loading progress of a specific instance
+    /// </summary>
+    public class NetworkedLoadingProgressTracker: NetworkBehaviour
+    {
+        NetworkVariable<float> m_Progress = new NetworkVariable<float>(0, NetworkVariableReadPermission.Everyone, NetworkVariableWritePermission.Owner);
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/NetworkedLoadingProgressTracker.cs:10]: Would ServerRPCs + local state be simpler here?

Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/NetworkedLoadingProgressTracker.cs:10]: Or user player + netvar in player for tracking this per client?

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/LoadingProgressManager.cs [hunk 4]
```diff
@@ -106,8 +111,15 @@ void RemoveTracker(ulong clientId)
                 var tracker = ProgressTrackers[clientId];
                 ProgressTrackers.Remove(clientId);
                 tracker.NetworkObject.Despawn();
-                UpdateTrackersClientRpc();
+                // This makes sure that clients received the Despawn message before the RPC.
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/LoadingProgressManager.cs:114]: This only makes sure RPCs are sent in different messages. They could be received in the same update client side.
Plus if you're sending messages reliable ordered, they should keep the same order client side.
What's the issue you saw here?

Comment (LPLafontaineB) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/LoadingProgressManager.cs:114]: Clients would be executing this RPC, but they wouldn't have despawned the networkObjects before executing it.
But good point, there is no guarantee of during which frame they are received client-side. I have some alternative solutions in mind that would be better

File: Assets/BossRoom/Prefabs/NetworkingManager.prefab [hunk 5]
```diff
@@ -218,6 +226,11 @@ MonoBehaviour:
       SourcePrefabToOverride: {fileID: 0}
       SourceHashToOverride: 0
       OverridingTargetPrefab: {fileID: 0}
+    - Override: 0
```
Comment (SamuelBellomo) [inline Assets/BossRoom/Prefabs/NetworkingManager.prefab:229]: after discussion, check this shouldn't be here

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/ClientLoadingScreen.cs [hunk 6]
```diff
@@ -63,26 +82,88 @@ public void StopLoadingScreen()
...
+
+        protected virtual void UpdateProgressBars(bool isInitializing)
+        {
+            foreach (var progressTracker in m_LoadingProgressManager.ProgressTrackers)
+            {
+                var clientId = progressTracker.Key;
+                var progress = progressTracker.Value.Progress;
+                if (clientId != NetworkManager.Singleton.LocalClientId)
+                {
+                    if (!m_ClientIdToProgressBarsIndex.ContainsKey(clientId))
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/ClientLoadingScreen.cs:133]: double if should be an &&

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/LoadingProgressManager.cs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+                m_LocalLoadOperation = value;
+            }
+        }
+
+        AsyncOperation m_LocalLoadOperation;
+
+        float m_LocalProgress;
+        public event Action onTrackersUpdated;
+
+        public float LocalProgress
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/LoadingProgressManager.cs:33]: Let's add some comments on public properties

Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/LoadingProgressManager.cs:33]: especially why here (offline vs online progress)

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/NetworkedLoadingProgressTracker.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+using Unity.Netcode;
+
+namespace Unity.Multiplayer.Samples.Utilities
+{
+    /// <summary>
+    /// Simple object that keeps track of the scene loading progress of a specific instance
+    /// </summary>
+    public class NetworkedLoadingProgressTracker: NetworkBehaviour
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/NetworkedLoadingProgressTracker.cs:8]: Our overall discussion over using owner netvars for client state tracking should be documented. We have some doc on client driven stuff, let's add something there, how it needs a NO per player for it to work.

**Review by fernando-cortez — CHANGES_REQUESTED** [review]
I'm getting a `Failed to spawn NetworkObject for Hash 3997454455` when hosting with editor and joining as a built client. This is with a fresh project.

**Review by SamuelBellomo — APPROVED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Review by fernando-cortez — APPROVED** [review]

**Pull request merged**